File: app/indexer/htmlmap.js
// Message id -> media path, recovered from an HTML export.
//
// A port of tools/tgindex/htmlmap.py. Telegram names sticker files after the
// sticker's own filename, so every send collides and gets a " (N)" suffix
// numbered per export run: the same sticker is "sticker (5).webp" in one
// export and "sticker (12).webp" in the next. An HTML export already on disk
// is therefore a correct map from message to bytes, and reading it lets a
// fresh JSON export reuse files instead of needing them downloaded again.
//
// A regex scan rather than a parser: the files are huge and machine-written.

import { MEDIA_DIRS } from './model.js';

const MESSAGE_RE = /<div class="message[^"]*"\s+id="message(-?\d+)"/g;

const escape = (text) => text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const ASSET_RE = new RegExp(
  `(?:href|src)="((?:${MEDIA_DIRS.map(escape).join('|')})/[^"]+)"`, 'gi');

const THUMB_SUFFIXES = ['_thumb.jpg', '_thumb.png', '_thumb.webp'];

function isThumb(path) {
  const lowered = path.toLowerCase();
  if (THUMB_SUFFIXES.some((suffix) => lowered.endsWith(suffix))) return true;
  // "sticker (2)_thumb (3).webp" -- a de-duplicated thumbnail.
  return /_thumb(?: \(\d+\))?\.\w+$/.test(lowered);
}

const NAMED = { amp: '&', lt: '<', gt: '>', quot: '"', apos: "'", nbsp: ' ' };

/** The HTML entities Telegram's exporter actually writes, as html.unescape. */
function unescapeHTML(text) {
  return text.replace(/&(#x[0-9a-f]+|#\d+|[a-z]+);/gi, (whole, body) => {
    if (body[0] === '#') {
      const code = body[1] === 'x' || body[1] === 'X'
        ? Number.parseInt(body.slice(2), 16)
        : Number.parseInt(body.slice(1), 10);
      try { return String.fromCodePoint(code); } catch { return whole; }
    }
    return NAMED[body.toLowerCase()] ?? whole;
  });
}

const utf8 = new TextDecoder('utf-8');

/** Percent-decoding as urllib.parse.unquote: bad bytes become U+FFFD. */
function unquote(text) {
  return text.replace(/(?:%[0-9a-f]{2})+/gi, (run) => {
    const bytes = run.match(/%[0-9a-f]{2}/gi).map((h) => Number.parseInt(h.slice(1), 16));
    return utf8.decode(Uint8Array.from(bytes));
  });
}

const clean = (raw) => unquote(unescapeHTML(raw));

/**
 * Add one HTML page's message -> media entries to `out`.
 *
 * One page at a time on purpose: a big export runs to hundreds of pages, and
 * holding them all in memory at once is how a browser tab runs out of it.
 * Earlier pages win, matching the order the Python indexer reads them in.
 */
export function addPageToMap(text, out) {
  const marks = [...text.matchAll(MESSAGE_RE)].map((m) => [m.index, Number(m[1])]);
  for (let i = 0; i < marks.length; i += 1) {
    const [start, id] = marks[i];
    const end = i + 1 < marks.length ? marks[i + 1][0] : text.length;
    const block = text.slice(start, end);

    let src = null;
    let thumb = null;
    for (const asset of block.matchAll(ASSET_RE)) {
      const candidate = clean(asset[1]);
      if (isThumb(candidate)) thumb ??= candidate;
      else src ??= candidate;
    }

    if ((src || thumb) && !out.has(id)) {
      const entry = {};
      if (src) entry.src = src;
      if (thumb) entry.thumb = thumb;
      out.set(id, entry);
    }
  }
  return out;
}

/** Map message id -> { src, thumb } for a list of page texts. */
export function parseMediaMap(pages) {
  const out = new Map();
  for (const text of pages) addPageToMap(text, out);
  return out;
}

File: tools/tgindex/htmlmap.py
"""Extract a message-id -> media-path map from an HTML export.

Why this exists: Telegram names sticker files after the sticker's own filename
attribute, which is literally ``sticker.webp`` for most packs.  Every send
collides on write and gets a ` (N)` suffix, and the numbering is assigned in
export order *per run*.  So the same sticker is `sticker (5).webp` in one export
and `sticker (12).webp` in the next, and a JSON export dropped into a folder
that already holds media will re-download rather than reuse.

That makes an already-downloaded HTML export folder valuable on its own: it is a
complete, correct mapping from message id to the bytes on disk.  This module
reads it so a fresh JSON export can supply the structure while an older (perhaps
hand-pruned) HTML folder supplies the files -- no re-download, no lost curation.

Deliberately a regex scan, not a real parser: these files run to hundreds of MB
and the markup is machine-generated and rigidly uniform.
"""

from __future__ import annotations

import html
import re
import urllib.parse
from pathlib import Path

from .model import MEDIA_DIRS

# <div class="message default clearfix" id="message171376">
MESSAGE_RE = re.compile(r'<div class="message[^"]*"\s+id="message(-?\d+)"')

# href/src pointing into one of the export's media folders.
_MEDIA_DIR_ALT = "|".join(re.escape(d) for d in MEDIA_DIRS)
ASSET_RE = re.compile(
    r'(?:href|src)="((?:%s)/[^"]+)"' % _MEDIA_DIR_ALT,
    re.IGNORECASE,
)

# Telegram writes thumbnails next to the real file under two different
# conventions depending on which exporter produced them.
THUMB_SUFFIXES = ("_thumb.jpg", "_thumb.png", "_thumb.webp")


def _is_thumb(path: str) -> bool:
    lowered = path.lower()
    if any(lowered.endswith(s) for s in THUMB_SUFFIXES):
        return True
    # `sticker (2)_thumb (3).webp` -- a de-duplicated thumbnail.
    return re.search(r"_thumb(?: \(\d+\))?\.\w+$", lowered) is not None


def _clean(raw: str) -> str:
    """Turn an href into a path relative to the export folder."""
    return urllib.parse.unquote(html.unescape(raw))


def parse_media_map(html_files: list[Path]) -> dict[int, dict[str, str]]:
    """Map message id -> {"src": path, "thumb": path} for one export folder.

    Paths are relative to the export folder, matching how result.json stores
    them.  Messages with no media are absent from the map.
    """
    out: dict[int, dict[str, str]] = {}

    for path in html_files:
        try:
            text = path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue

        # Split on message boundaries so each asset is attributed to the message
        # that contains it.  finditer + slice avoids materialising a copy per
        # message the way re.split would.
        marks = [(m.start(), int(m.group(1))) for m in MESSAGE_RE.finditer(text)]
        for i, (start, msg_id) in enumerate(marks):
            end = marks[i + 1][0] if i + 1 < len(marks) else len(text)
            block = text[start:end]

            src: str | None = None
            thumb: str | None = None
            for asset in ASSET_RE.finditer(block):
                candidate = _clean(asset.group(1))
                if _is_thumb(candidate):
                    thumb = thumb or candidate
                else:
                    src = src or candidate

            if src or thumb:
                entry: dict[str, str] = {}
                if src:
                    entry["src"] = src
                if thumb:
                    entry["thumb"] = thumb
                # Later files win only if they add something; a message never
                # legitimately appears in two HTML pages of the same export.
                out.setdefault(msg_id, entry)

    return out
